File: .github/scripts/test-mcp-apps-enabled-profile.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "fileutils"
require "find"
require "json"
require "open3"
require "tempfile"
require "timeout"
require "tmpdir"

CommandResult = Struct.new(:stdout, :stderr, :status, :timed_out, keyword_init: true)

source_root = File.expand_path("../..", __dir__)
fixture_root = File.join(source_root, ".github/fixtures/profiles/mcp-apps-enabled")
validator = File.join(source_root, "template/.github/scripts/validate_skill_repository.py")
expected_files = %w[
  INTERFACES.md
  MCP_APPS.md
  MCP_INTERFACE.md
  RUNTIME.md
  SKILL.md
  docs/mcp-transports.md
  mcp/apps/host_bridge.mjs
  mcp/apps/result.html
  mcp/server.mjs
  package.json
  src/text_stats.mjs
  tests/test_mcp_apps.mjs
].sort.freeze

run_command = lambda do |*command, chdir:, timeout_seconds:, env: {}|
  stdout_file = Tempfile.new("mcp-apps-fixture-stdout")
  stderr_file = Tempfile.new("mcp-apps-fixture-stderr")
  status = nil
  timed_out = false
  pid = nil

  begin
    pid = Process.spawn(
      env,
      *command,
      chdir: chdir,
      in: File::NULL,
      out: stdout_file.path,
      err: stderr_file.path,
      pgroup: true
    )
    begin
      Timeout.timeout(timeout_seconds) do
        _waited_pid, status = Process.wait2(pid)
      end
    rescue Timeout::Error
      timed_out = true
      begin
        Process.kill("TERM", -pid)
      rescue Errno::ESRCH
        nil
      end
      begin
        Timeout.timeout(2) { _waited_pid, status = Process.wait2(pid) }
      rescue Timeout::Error
        begin
          Process.kill("KILL", -pid)
        rescue Errno::ESRCH
          nil
        end
        _waited_pid, status = Process.wait2(pid)
      rescue Errno::ECHILD
        nil
      end
    end
  rescue Errno::ENOENT => error
    return CommandResult.new(stdout: "", stderr: "#{error.message}\n", status: nil, timed_out: false)
  ensure
    stdout_file.flush
    stderr_file.flush
  end

  CommandResult.new(
    stdout: File.binread(stdout_file.path),
    stderr: File.binread(stderr_file.path),
    status: status,
    timed_out: timed_out
  )
ensure
  stdout_file&.close!
  stderr_file&.close!
end

failures = []
actual_files = Find.find(fixture_root).filter_map do |path|
  next if path == fixture_root || File.directory?(path)

  path.delete_prefix("#{fixture_root}/")
end.sort
failures << "mcp-apps-enabled layout drift: #{actual_files.inspect}" unless actual_files == expected_files

begin
  manifest = JSON.parse(File.read(File.join(fixture_root, "package.json"), encoding: "UTF-8"))
  dependencies = manifest.fetch("dependencies")
  {
    "@modelcontextprotocol/server" => "2.0.0",
    "@modelcontextprotocol/client" => "2.0.0",
    "zod" => "4.1.13"
  }.each do |package, version|
    failures << "mcp-apps-enabled: expected exact #{package}=#{version} pin" unless dependencies[package] == version
  end
rescue JSON::ParserError, KeyError => error
  failures << "mcp-apps-enabled package.json: #{error.message}"
end

runtime = File.read(File.join(fixture_root, "RUNTIME.md"), encoding: "UTF-8")
failures << "mcp-apps-enabled: missing core 2026-07-28 selection" unless runtime.include?("| Supported protocol revisions | `2026-07-28` |")
failures << "mcp-apps-enabled: missing exact Apps extension selection" unless runtime.include?("| Optional MCP extensions | io.modelcontextprotocol/ui |")
apps = File.read(File.join(fixture_root, "MCP_APPS.md"), encoding: "UTF-8")
failures << "mcp-apps-enabled: missing Apps 2026-01-26 revision" unless apps.include?("Extension specification revision: 2026-01-26")
failures << "mcp-apps-enabled: must remain independent of browser-interface" if File.exist?(File.join(fixture_root, "WEB_INTERFACE.md"))

unless failures.empty?
  failures.each { |failure| warn failure }
  exit 1
end

Dir.mktmpdir("mcp-apps-profile") do |directory|
  FileUtils.cp_r("#{fixture_root}/.", directory)
  run_command.call("git", "init", "--quiet", chdir: directory, timeout_seconds: 10)
  run_command.call("git", "add", ".", chdir: directory, timeout_seconds: 10)

  validation = run_command.call(
    "python3", validator, directory,
    chdir: source_root,
    timeout_seconds: 60,
    env: { "RUBYOPT" => nil }
  )
  unless validation.status&.success? && !validation.timed_out
    failures << "mcp-apps-enabled contract validation failed: stdout=#{validation.stdout.inspect}, stderr=#{validation.stderr.inspect}, timed_out=#{validation.timed_out}"
  end

  install = run_command.call(
    "npm", "install", "--ignore-scripts", "--no-audit", "--no-fund",
    chdir: directory,
    timeout_seconds: 180
  )
  unless install.status&.success? && !install.timed_out
    failures << "mcp-apps-enabled npm install failed: stdout=#{install.stdout.inspect}, stderr=#{install.stderr.inspect}, timed_out=#{install.timed_out}"
  end

  if install.status&.success?
    %w[check test].each do |script|
      result = run_command.call("npm", "run", script, chdir: directory, timeout_seconds: 120)
      unless result.status&.success? && !result.timed_out
        failures << "mcp-apps-enabled npm run #{script} failed: stdout=#{result.stdout.inspect}, stderr=#{result.stderr.inspect}, timed_out=#{result.timed_out}"
      end
    end
  end
end

Dir.mktmpdir("invalid-mcp-apps-profile") do |directory|
  FileUtils.cp_r("#{fixture_root}/.", directory)
  File.delete(File.join(directory, "MCP_APPS.md"))
  run_command.call("git", "init", "--quiet", chdir: directory, timeout_seconds: 10)
  run_command.call("git", "add", ".", chdir: directory, timeout_seconds: 10)

  validation = run_command.call(
    "python3", validator, directory,
    chdir: source_root,
    timeout_seconds: 60,
    env: { "RUBYOPT" => nil }
  )
  if validation.timed_out || validation.status&.success?
    failures << "mcp-apps-enabled negative contract: selected Apps without MCP_APPS.md must fail"
  end
end

unless failures.empty?
  failures.each { |failure| warn failure }
  exit 1
end

puts "MCP Apps executable fixture tests passed."

File: .github/scripts/test_mcp_apps_enabled_profile.py
#!/usr/bin/env python3
"""Exercise the MCP Apps 2026-01-26 fixture without Ruby."""

from __future__ import annotations

import json
import os
import shutil
import signal
import subprocess
import sys
import tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
FIXTURE = ROOT / ".github/fixtures/profiles/mcp-apps-enabled"
VALIDATOR = ROOT / "template/.github/scripts/validate_skill_repository.py"
EXPECTED = sorted([
    "INTERFACES.md", "MCP_APPS.md", "MCP_INTERFACE.md", "RUNTIME.md", "SKILL.md",
    "docs/mcp-transports.md", "mcp/apps/host_bridge.mjs", "mcp/apps/result.html",
    "mcp/server.mjs", "package.json", "src/text_stats.mjs", "tests/test_mcp_apps.mjs",
])
PINS = {
    "@modelcontextprotocol/server": "2.0.0",
    "@modelcontextprotocol/client": "2.0.0",
    "zod": "4.1.13",
}


def environment() -> dict[str, str]:
    env = os.environ.copy()
    for key in ("RUBYOPT", "PYTHONPATH", "GIT_DIR", "GIT_INDEX_FILE", "GIT_WORK_TREE"):
        env.pop(key, None)
    env["PYTHONUTF8"] = "1"
    return env


def run(command: list[str], *, cwd: Path, timeout: float) -> subprocess.CompletedProcess[str]:
    process = subprocess.Popen(
        command, cwd=cwd, env=environment(), stdin=subprocess.DEVNULL,
        stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, encoding="utf-8",
        start_new_session=(os.name != "nt"),
    )
    try:
        stdout, stderr = process.communicate(timeout=timeout)
        return subprocess.CompletedProcess(command, process.returncode, stdout, stderr)
    except subprocess.TimeoutExpired:
        if os.name != "nt":
            try:
                os.killpg(process.pid, signal.SIGTERM)
            except ProcessLookupError:
                pass
        else:
            process.terminate()
        try:
            stdout, stderr = process.communicate(timeout=2)
        except subprocess.TimeoutExpired:
            if os.name != "nt":
                try:
                    os.killpg(process.pid, signal.SIGKILL)
                except ProcessLookupError:
                    pass
            else:
                process.kill()
            stdout, stderr = process.communicate()
        return subprocess.CompletedProcess(command, 124, stdout, stderr + "command timed out\n")


def init_git(directory: Path) -> None:
    for command in (["git", "init", "--quiet"], ["git", "add", "."]):
        result = run(command, cwd=directory, timeout=10)
        if result.returncode != 0:
            raise RuntimeError(f"{' '.join(command)} failed: {result.stderr}")


def main() -> int:
    failures: list[str] = []
    actual = sorted(
        path.relative_to(FIXTURE).as_posix()
        for path in FIXTURE.rglob("*") if not path.is_dir()
    )
    if actual != EXPECTED:
        failures.append(f"mcp-apps-enabled layout drift: {actual!r}")

    try:
        manifest = json.loads((FIXTURE / "package.json").read_text(encoding="utf-8"))
        dependencies = manifest["dependencies"]
        for package, version in PINS.items():
            if dependencies.get(package) != version:
                failures.append(f"mcp-apps-enabled: expected exact {package}={version} pin")
    except (OSError, KeyError, json.JSONDecodeError) as exc:
        failures.append(f"mcp-apps-enabled package.json: {exc}")

    runtime = (FIXTURE / "RUNTIME.md").read_text(encoding="utf-8")
    if "| Supported protocol revisions | `2026-07-28` |" not in runtime:
        failures.append("mcp-apps-enabled: missing core 2026-07-28 selection")
    if "| Optional MCP extensions | io.modelcontextprotocol/ui |" not in runtime:
        failures.append("mcp-apps-enabled: missing exact Apps extension selection")
    apps = (FIXTURE / "MCP_APPS.md").read_text(encoding="utf-8")
    if "Extension specification revision: 2026-01-26" not in apps:
        failures.append("mcp-apps-enabled: missing Apps 2026-01-26 revision")
    if (FIXTURE / "WEB_INTERFACE.md").exists():
        failures.append("mcp-apps-enabled: must remain independent of browser-interface")

    if failures:
        for failure in failures:
            print(failure, file=sys.stderr)
        return 1

    with tempfile.TemporaryDirectory(prefix="mcp-apps-profile") as temporary:
        directory = Path(temporary)
        shutil.copytree(FIXTURE, directory, dirs_exist_ok=True, symlinks=True)
        init_git(directory)
        validation = run([sys.executable, str(VALIDATOR), str(directory)], cwd=ROOT, timeout=60)
        if validation.returncode != 0:
            failures.append(
                f"mcp-apps-enabled contract validation failed: stdout={validation.stdout!r}, stderr={validation.stderr!r}"
            )
        install = run(
            ["npm", "install", "--ignore-scripts", "--no-audit", "--no-fund"],
            cwd=directory, timeout=180,
        )
        if install.returncode != 0:
            failures.append(f"mcp-apps-enabled npm install failed: {install.stderr!r}")
        else:
            for script in ("check", "test"):
                result = run(["npm", "run", script], cwd=directory, timeout=120)
                if result.returncode != 0:
                    failures.append(
                        f"mcp-apps-enabled npm run {script} failed: stdout={result.stdout!r}, stderr={result.stderr!r}"
                    )

    with tempfile.TemporaryDirectory(prefix="invalid-mcp-apps-profile") as temporary:
        directory = Path(temporary)
        shutil.copytree(FIXTURE, directory, dirs_exist_ok=True, symlinks=True)
        (directory / "MCP_APPS.md").unlink()
        init_git(directory)
        validation = run([sys.executable, str(VALIDATOR), str(directory)], cwd=ROOT, timeout=60)
        if validation.returncode == 0:
            failures.append("mcp-apps-enabled negative contract: selected Apps without MCP_APPS.md must fail")

    if failures:
        for failure in failures:
            print(failure, file=sys.stderr)
        return 1
    print("MCP Apps executable fixture tests passed.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
